topdown/http.go

```diff
@@ -569,6 +680,9 @@ func createHTTPRequest(bctx BuiltinContext, obj ast.Object) (*http.Request, *htt
 		tlsConfig.ServerName = tlsServerName
 	}
 
+	// cache the client and request
+	clientCache.Insert(obj, &httpClientCacheValue{client: client, req: req, files: files, envVars: envVars})
+
```

Comment:
This means we'll create and cache one client per obj, right?

So if I make 12 requests with some header that make 12 http.send calls to some other service including that header value, it'll cache those 12 clients. Will they ever get evicted or the cache get reset?

I was wondering about the rationale behind this change - I had assumed we'd get done connection reuse out of this, perhaps, but that wouldn't be possible if the requests (assuming to the same server) each have their own client.

Reply:
> This means we'll create and cache one client per obj, right?

Yes.


> So if I make 12 requests with some header that make 12 http.send calls to some other service including that header value, it'll cache those 12 clients. Will they ever get evicted or the cache get reset?

If the header values are the same they get cached. If the underlying file contents or env variables have changed we update the cache. Other than that there is no reset mechanism or memory limit on this cache. Maybe it makes sense to have one.


> I was wondering about the rationale behind this change - I had assumed we'd get done connection reuse out of this, perhaps, but that wouldn't be possible if the requests (assuming to the same server) each have their own client.

By this you mean, the request object varies in each request to the same server ? If yes, then we will cache each client.

Reply:
> Other than that there is no reset mechanism or memory limit on this cache. Maybe it makes sense to have one.

So if I were to include `time.now_ns` in the request, I'd eventually run out of memory?

Reply:
> So if I were to include time.now_ns in the request, I'd eventually run out of memory?

Yes. We need to set a reasonable limit for this.

Reply:
A compromise might be to cache the client using the host only... There's unlikely to be that much variance in target hosts... 🤔